File: project/unigram-backend/AAC.py
from typing import List, Dict, Any
import requests
from datetime import datetime

class Keyword:
    def __init__(self, name: str):
        self.name = name

class Category:
    def __init__(self, name: str, label: str, tags: List[str], keywords: List[str], children: List['Category']):
        self.name = name
        self.label = label
        self.tags = tags
        self.keywords = keywords
        self.children = children

class Pictogram:
    def __init__(self, _id: int, created: datetime, tags: List[str], keywords: List[Keyword], url: str = None, categories: List[str] = []):
        self._id = _id
        self.created = created
        self.tags = tags
        self.keywords = keywords
        self.url = url
        self.categories = categories

class AAC:
    def __init__(self, language: str):
        self.language = language
        
    def convert_sentence(self, sentence: str) -> str:
        res = requests.get(f"https://api.arasaac.org/v1/phrases/flex/es/{sentence}")
        data = res.json()
        #if contains error
        if "error" in data:
            return False
        return data['msg']
        

    def get_keywords(self) -> List[Keyword]:
        res = requests.get(f"https://api.arasaac.org/api/keywords/{self.language}")
        data = res.json()
        result = [Keyword(keyword['name']) for keyword in data]
        return result

    def fetch_categories_recursive(self, language: str) -> List[Category]:
        res = requests.get(f"https://privateapi.arasaac.org/api/categories/{language}")
        data = res.json()
        if "error" in data or len(data.get("data", [])) == 0:
            return []

        def build_category(category_name: str, category_data: Dict[str, Any]) -> Category:
            return Category(category_name, category_data['text'], category_data['tags'], category_data['keywords'], 
                            [build_category(child_name, child_data) for child_name, child_data in category_data.get('children', {}).items()])

        return [build_category(category_name, data['data'][category_name]) for category_name in data['data']]

    def get_categories(self) -> List[Category]:
        return self.fetch_categories_recursive(self.language)

    def search_keyword(self, keyword: str, normal: bool = True) -> List[Pictogram]:
        if not keyword:
            return []
        res = requests.get(f"https://api.arasaac.org/api/pictograms/{self.language}/{'search' if normal else 'bestsearch'}/{keyword}")
        data = res.json()
        if not data:
            return []
        if "error" in data:
            return []
        result = []
        for pictogram in data:
            keywords = []
            for keyword_obj in pictogram.get('keywords', []):
                if 'name' in keyword_obj:
                    keywords.append(Keyword(keyword_obj['name']))
            result.append(Pictogram(pictogram['_id'], datetime.fromisoformat(pictogram['created']), pictogram['tags'], keywords))
        return result

    def get_info_from_id(self, _id: int):
        res = requests.get(f"https://api.arasaac.org/api/pictograms/{self.language}/{_id}")
        data = res.json()
        
        # Check if 'keywords' key exists in data
        if 'keywords' in data:
            keywords = [Keyword(keyword_obj['name']) for keyword_obj in data['keywords'] if 'name' in keyword_obj]
        else:
            keywords = []
        
        result = Pictogram(
            data['_id'], 
            datetime.fromisoformat(data['created']), 
            data.get('tags', []), 
            keywords,
            None,
            data.get('categories', [])
        )
        return result

    def get_image_from_id(self, _id: int, color: bool = True, skin: str = "white", hair: str = "red") -> Pictogram:
        res = requests.get(f"https://api.arasaac.org/v1/pictograms/{_id}?plural=false&color=true&skin={skin}&hair={hair}&url=true&download=false")
        data = res.json()
        if 'image' not in data:
            raise KeyError("Image URL not found in response")
        image_url = data['image']
        pictogram = self.get_info_from_id(_id)
        pictogram.url = image_url
        return pictogram

File: src/lib/AAC.ts
/**
 * Enum representing different hair colors.
 */
export enum HairColor {
    BLONDE = "blonde",
    BROWN = "brown",
    DARK_BROWN = "darkBrown",
    GRAY = "gray",
    DARK_DRAY = "darkGray",
    RED = "red",
    BLACK = "black",
}

/**
 * Enum representing different skin colors.
 */
export enum SkinColor {
    WHITE = "white",
    BLACK = "black",
    ASSIAN = "assian",
    MULATTO = "mulatto",
    AZTEC = "aztec",
}

/**
 * Type representing a keyword.
 */
export type Keyword = {
    name: string;
}

/**
 * Type representing a category.
 */
export type Category = {
    name: string;
    label: string;
    tags: string[];
    keywords: string[];
    children: Category[];
}

/**
 * Interface representing a pictogram.
 */
export interface Pictogram {
    word?: string;
    schematic: boolean;
    sex: boolean;
    violence: boolean;
    aac: boolean;
    aacColor: boolean;
    skin: boolean;
    hair: boolean;
    categories: string[];
    synsets: string[];
    tags: string[];
    _id: number;
    desc: string;
    url: string;
}

/**
 * Class representing AAC functionalities.
 */
export class AAC {
    language: string;

    /**
     * Constructs an instance of AAC.
     * @param language - The language to be used.
     */
    constructor(language: string) {
        this.language = language;
    }

    /**
     * Converts a hair color to its corresponding hex value.
     * @param color - The hair color.
     * @returns The hex value of the hair color.
     */
    static hairColorToHex = (color: HairColor): string => {
        switch (color) {
            case HairColor.BLONDE:
                return "FDD700";
            case HairColor.BROWN:
                return "A65E26";
            case HairColor.DARK_BROWN:
                return "6A2703";
            case HairColor.GRAY:
                return "EFEFEF";
            case HairColor.DARK_DRAY:
                return "AAABAB";
            case HairColor.RED:
                return "ED4120";
            case HairColor.BLACK:
                return "020100";
            default:
                return "#FDD700";
        }
    }

    /**
     * Converts a skin color to its corresponding hex value.
     * @param color - The skin color.
     * @returns The hex value of the skin color.
     */
    static skinColorToHex = (color: SkinColor): string => {
        switch (color) {
            case SkinColor.WHITE:
                return "F5E5DE";
            case SkinColor.BLACK:
                return "A65C17";
            case SkinColor.ASSIAN:
                return "F4ECAD";
            case SkinColor.MULATTO:
                return "E3AB72";
            case SkinColor.AZTEC:
                return "CF9D7C";
            default:
                return "CF9D7C";
        }
    }

    /**
     * Fetches keywords from the API.
     * @returns A promise that resolves to an array of keywords.
     */
    getKeywords: () => Promise<Keyword[]> = async () => {
        let res = await fetch(`https://api.arasaac.org/api/keywords/${this.language}`, {
            headers: {
                "Accept": "*/*",
            },
            method: "GET"
        });
        let data = await res.json();
        //convert data to Keyword
        const result: Keyword[] = data.map((keyword: any) => {
            return {
                name: keyword.keyword
            }
        });
        return result;
    }

    /**
     * Recursively fetches categories from the API.
     * @param language - The language to be used.
     * @returns A promise that resolves to an array of categories.
     */
    fetchCategoriesRecursive = async (language: string): Promise<Category[]> => {
        let res = await fetch(`https://privateapi.arasaac.org/api/categories/${language}`, {
            headers: {
                Accept: "*/*",
            },
            method: "GET",
        });
        let data = await res.json();
        if (data["error"] || data["data"].length === 0) return [];
        
        const buildCategory = (categoryName:string, categoryData: any): Category => {
            return {
                name: categoryName,
                label: categoryData.text,
                tags: categoryData.tags,
                keywords: categoryData.keywords,
                children: (categoryData.children) ? Object.keys(categoryData.children).map((childName: any) => buildCategory(childName,categoryData.children[childName])) : []
            };
        };
    
        return Object.keys(data["data"]).map((categoryName: any) => buildCategory(categoryName,data["data"][categoryName]));
    };
    
    /**
     * Fetches categories from the API.
     * @returns A promise that resolves to an array of categories.
     */
    getCategories: () => Promise<Category[]> = async () => {
        return await this.fetchCategoriesRecursive(this.language);
    };

    /**
     * Searches for pictograms by keyword.
     * @param keyword - The keyword to search for.
     * @param normal - Whether to use normal search or best search.
     * @returns A promise that resolves to an array of pictograms.
     */
    searchKeyword: (keyword: string, normal: boolean) => Promise<Pictogram[]> = async (keyword = '', normal = true) => {
        if (keyword === '') return [];
        let res = await fetch(`https://api.arasaac.org/api/pictograms/${this.language}/${(normal)?"search":"bestsearch"}/${keyword}`, {
            "headers": {
                "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            },
            "method": "GET",
        })
        let data = await res.json();
        if (data.length === 0) return [];
        if (data["error"]) return [];
        const result: Pictogram[] = data.map((pictogram: any) => {
            return {
                _id: pictogram._id,
                created: pictogram.created,
                tags: pictogram.tags,
                keywords: pictogram.keywords
            }
        });
        return result;
    }

    /**
     * Fetches pictogram information by ID.
     * @param id - The ID of the pictogram.
     * @returns A promise that resolves to a pictogram.
     */
    getInfoFromId: (id: number) => Promise<Pictogram> = async (id = 0) => {
        let res = await fetch(`https://api.arasaac.org/api/pictograms/${this.language}/${id}`, {
            "headers": {
                "Accept": "*/*",
            },
            "method": "GET",
        })
        let data = await res.json();
        return data as Pictogram;
    }

    /**
     * Fetches pictogram image by ID.
     * @param id - The ID of the pictogram.
     * @param color - Whether to use color.
     * @param skin - The skin color.
     * @param hair - The hair color.
     * @returns A promise that resolves to a pictogram with the image URL.
     */
    getImageFromId: (id: number, color: boolean, skin: SkinColor, hair: HairColor) => Promise<Pictogram> = async (id = 0, color = true, skin = SkinColor.ASSIAN, hair = HairColor.RED) => {
        let res = await fetch(`https://api.arasaac.org/v1/pictograms/${id}?plural=false&color=${color}&skin=${skin.toString()}&hair=${hair.toString()}&url=true&download=false`, {
            "headers": {
                "Accept": "*/*",
            },
            "method": "GET",
        });
        let imageUrl = (await res.json()).image;
        let pictogram: Pictogram = await this.getInfoFromId(id);
        pictogram.url = imageUrl;
        return pictogram;
    }
}
